---
<h1 align="center"><strong>Cálculo de dependencia cruzada para todos los países</strong></h1>
<h4 align="center"><strong>Manuel Alejandro Hidalgo y Jorge Díaz Lanchas</strong></h4>
<h4 align="center"><strong>Fundación Real Instituto Elcano</strong></h4>

---

# Esquema para Índice de Seguridad Económica - Real Instituto Elcano

## 1. Introducción y Marco Conceptual

- **Objetivo**: Desarrollar un índice que cuantifique la seguridad económica de los países
- **Definición**: La seguridad económica como la capacidad de un país para resistir disrupciones en sus cadenas de suministro y comercio internacional
- **Relevancia**: Contexto actual de fragmentación geoeconómica y tensiones comerciales

## 2. Metodología

### 2.1 Fuentes de Datos
- Base de datos International Trade and Production Database (ITP)
- Datos comerciales bilaterales por industria (año 2019)
- Otros indicadores macroeconómicos complementarios

### 2.2 Procesamiento de Datos
```python
# Procesamiento y carga de datos ITP
itp2019, codigos_countries = procesar_datos_itp()
```

### 2.3 Creación de Matrices de Comercio
```python
# Generación de matrices bilaterales por industria
matrices_comercio = crear_matriz_comercio(data.groupby('industry_descr'), codigos_paises)
```

### 2.4 Limpieza y Filtrado
```python
# Eliminar relaciones comerciales insignificantes
mat_clean = eliminar_filas_columnas_cero(mat, threshold_pct=0.05)
```

### 2.5 Cálculo de Dependencias Económicas
```python
# Cálculo de dependencias directas e indirectas
results = analyze_dependencies(X, country_names)
```

## 3. Componentes del Índice

### 3.1 Dependencia Directa
- Medición de la dependencia inmediata entre países
- Fórmula: $D_{ij} = \frac{X_{ji}}{∑_k X_{ki}}$, donde $X_{ji}$ es el comercio del país j al país i

### 3.2 Dependencia Indirecta
- Medición de dependencias a través de cadenas de suministro
- Incorporación de países intermediarios en las relaciones comerciales
- Análisis de caminos hasta longitud 5

In [1]:
# MEJOR - con validación y logging
import os as _os
_threading_envs = {
    "OPENBLAS_NUM_THREADS": "1",
    "MKL_NUM_THREADS": "1", 
    "OMP_NUM_THREADS": "1",
    "NUMEXPR_NUM_THREADS": "1"
}
for env_var, value in _threading_envs.items():
    _os.environ.setdefault(env_var, value)
    print(f"✓ {env_var} = {value}")
    
import os
import gzip
import pandas as pd
import numpy as np
from numpy.linalg import inv
from tqdm import tqdm
import matplotlib.pyplot as plt
from pathlib import Path
from itertools import combinations
import torch

# Añadimos soporte esparso para futuros cálculos eficientes
from scipy import sparse  # <- nuevo

import dask.dataframe as dd
from concurrent.futures import ThreadPoolExecutor
from typing import List, Dict
import multiprocessing
from joblib import Parallel, delayed

import warnings
warnings.filterwarnings("ignore", category=UserWarning, module="joblib")


# (Opcional) detección segura de CuPy sin importarlo directamente (evita aviso del linter)
import importlib.util as _importlib_util
_cp_spec = _importlib_util.find_spec("cupy")
cp = None  # placeholder para evitar NameError si luego lo referenciamos
if _cp_spec is not None:
    # Solo importamos si realmente está instalado
    import importlib as _importlib
    cp = _importlib.import_module("cupy")



✓ OPENBLAS_NUM_THREADS = 1
✓ MKL_NUM_THREADS = 1
✓ OMP_NUM_THREADS = 1
✓ NUMEXPR_NUM_THREADS = 1


***No ejecutar este código a menos que se quiera comprimir***

In [ ]:
def comprimir_dividir_archivo(archivo_original, tamano_maximo=100, directorio_salida=None):
    # Asegúrate de que el archivo original existe
    archivo_original = Path(archivo_original)
    if not archivo_original.exists():
        raise FileNotFoundError(f"No se encuentra el archivo: {archivo_original}")
    
    # Si no se especifica directorio de salida, usar src/data/raw/ITP/
    if directorio_salida is None:
        # Obtener el directorio raíz del proyecto (donde está src/)
        proyecto_root = Path.cwd().parent.parent
        directorio_salida = proyecto_root /'data' / 'raw' / 'ITP' / 'ITPD_E_R03'
    else:
        directorio_salida = Path(directorio_salida)
    
    # Crear el directorio de salida si no existe
    directorio_salida.mkdir(parents=True, exist_ok=True)
    
    # Abre el archivo original en modo de lectura binaria
    with open(archivo_original, 'rb') as f_in:
        # Lee el contenido del archivo original
        contenido = f_in.read()
        
        # Determina el número de partes necesarias
        num_partes = (len(contenido) + tamano_maximo - 1) // tamano_maximo
        
        # Divide el contenido en partes y escribe cada parte comprimida
        for i in range(num_partes):
            parte = contenido[i * tamano_maximo: (i + 1) * tamano_maximo]
            archivo_salida = directorio_salida / f'ITPD_E_R03.csv.parte{i}.gz'
            with gzip.open(archivo_salida, 'wb') as f_out:
                f_out.write(parte)
            print(f"Parte {i} creada en: {archivo_salida}")

# Tamaño máximo por parte (1GB)
tamano_maximo = 810 * 1024 * 1024

try:
    # Ruta al archivo original
    archivo_original = Path(r"C:\Users\Usuario\Downloads\ITPDE_R03\ITPDE_R03.csv")
    
    # Comprimir y dividir el archivo original
    comprimir_dividir_archivo(archivo_original, tamano_maximo)
    print("Proceso completado con éxito")
except Exception as e:
    print(f"Error durante el proceso: {e}")

### ***Descomprimir, carga de datos y borrado de archivo***

La compresión se hace para poder trabajar con git sin porblemas de tamaño de ficheros.
Se descomprime, se importa y luego se borra el fichero descomprimido


# DEFINO EL AÑO

In [2]:
anio = 2022

In [3]:
"""
FASE 1: PREPARACIÓN Y CARGA DE DATOS
Este script procesa la base de datos International Trade and Production Database (ITP)
que viene dividida en múltiples archivos comprimidos, utilizando aceleración GPU
cuando está disponible.
"""

def procesar_datos_itp(year: int = anio):
    try:
        # Verificar si GPU está disponible (solo informativo en esta fase)
        gpu_disponible = torch.cuda.is_available()
        if gpu_disponible:
            print(f"GPU disponible: {torch.cuda.get_device_name(0)}")
        else:
            print("GPU no disponible, se usará CPU")

        # Definición de rutas usando Path y la estructura de tu proyecto
        try:
            base_path = Path(__file__).parent
        except NameError:  # Estamos en un notebook
            base_path = Path.cwd().parent.parent  # Asumiendo que el notebook está en /notebooks/

        source_directory = base_path / "data" / "raw" / "ITP" / "ITPD_E_R03"
        target_directory = base_path / "data" / "processed"
        target_filename = 'ITPD_E_R03.csv'

        # Imprimir las rutas para verificación
        print(f"Directorio fuente: {source_directory}")
        print(f"Directorio destino: {target_directory}")

        # Asegurar que los directorios existen
        target_directory.mkdir(parents=True, exist_ok=True)

        # Verificar que el directorio fuente existe
        if not source_directory.exists():
            raise FileNotFoundError(f"No se encuentra el directorio fuente: {source_directory}")

        # Listar archivos comprimidos
        chunk_filenames = sorted([
            f for f in os.listdir(source_directory)
            if f.startswith('ITPD_E_R03.csv.parte') and f.endswith('.gz')
        ])

        # Control de errores: verificar que existen archivos para procesar
        if not chunk_filenames:
            raise FileNotFoundError(f"No se encontraron archivos .gz en {source_directory}")

        # Construir la ruta completa para el archivo combinado
        target_filepath = target_directory / target_filename

        # Función para descomprimir un archivo en paralelo
        def descomprimir_archivo(chunk_filename):
            chunk_filepath = source_directory / chunk_filename
            with gzip.open(chunk_filepath, 'rb') as chunk_file:
                return chunk_file.read()

        print("Combinando archivos comprimidos...")
        with open(target_filepath, 'wb') as target_file:
            # Usar ThreadPoolExecutor para paralelizar la descompresión (I/O + gzip)
            with ThreadPoolExecutor(max_workers=os.cpu_count()) as executor:
                for data in tqdm(
                    executor.map(descomprimir_archivo, chunk_filenames),
                    total=len(chunk_filenames),
                    desc="Procesando archivos"
                ):
                    target_file.write(data)

                print("Leyendo archivo CSV...")

                # === columnas exactas que necesitas ===
                USECOLS = [
                    "exporter_iso3",
                    "importer_iso3",
                    "year",
                    "trade",
                    "industry_id",
                    "industry_descr",
                    "importer_name",
                    "exporter_name",
                ]
                DTYPES = {
                    "exporter_iso3": "string",
                    "importer_iso3": "string",
                    "year": "int32",
                    "trade": "float32",        # reduce memoria sin perder precisión práctica
                    "industry_id": "int32",
                    "industry_descr": "string",
                    "importer_name": "string",
                    "exporter_name": "string",
                }

                # Dask: lectura lazy del CSV combinado
                print("Usando Dask para procesamiento en paralelo (lectura filtrada)")
                dask_df = dd.read_csv(
                    target_filepath,
                    sep=",",
                    usecols=USECOLS,
                    dtype=DTYPES,
                    blocksize="128MB",
                    assume_missing=True,  # tolerante a nulos esporádicos
                )
                dask_df['year'].unique()
                print(f"Filtrando datos del año {year}...")
                dask_df = dask_df[dask_df["year"] == year]

                # Ejecutar el plan y materializar en pandas
                itp_year = dask_df.compute()

                # Tipos finales (por si Dask promovió algo)
                itp_year = itp_year.astype(DTYPES)

                # (Opcional) comprime memoria de las cadenas con 'category' si vas a agrupar mucho después
                # itp_year["exporter_iso3"]  = itp_year["exporter_iso3"].astype("category")
                # itp_year["importer_iso3"]  = itp_year["importer_iso3"].astype("category")
                # itp_year["industry_descr"] = itp_year["industry_descr"].astype("category")
                # itp_year["importer_name"]  = itp_year["importer_name"].astype("category")
                # itp_year["exporter_name"]  = itp_year["exporter_name"].astype("category")


        # Limpieza del temporal grande
        try:
            os.remove(target_filepath)
            print("Archivo temporal eliminado")
        except Exception as _e:
            print(f"Advertencia: no se pudo eliminar el temporal ({_e})")

        # 4) Países únicos importadores (GPU no aporta aquí; unique de pandas es muy rápido)
        #    Convertimos a category para memoria/velocidad y extraemos categorías ordenadas
        itp_year["importer_iso3"] = itp_year["importer_iso3"].astype("category")
        codigos_countries = list(itp_year["importer_iso3"].cat.categories)

        print(f"Total de países únicos encontrados: {len(codigos_countries)}")
        return itp_year, codigos_countries

    except Exception as e:
        print(f"Error durante el procesamiento: {e}")
        raise

if __name__ == "__main__":
    try:
        data, countries = procesar_datos_itp(year=anio)
        print("Procesamiento completado con éxito")
    except Exception as e:
        print(f"Error en la ejecución principal: {e}")


GPU disponible: NVIDIA GeForce RTX 4060 Laptop GPU
Directorio fuente: c:\Users\Usuario\Documents\Github\Seguridad Economica\data\raw\ITP\ITPD_E_R03
Directorio destino: c:\Users\Usuario\Documents\Github\Seguridad Economica\data\processed
Combinando archivos comprimidos...


Procesando archivos: 100%|██████████| 10/10 [00:17<00:00,  1.72s/it]


Leyendo archivo CSV...
Usando Dask para procesamiento en paralelo (lectura filtrada)
Filtrando datos del año 2022...
Archivo temporal eliminado
Total de países únicos encontrados: 236
Procesamiento completado con éxito


In [4]:
import pandas as pd
from pathlib import Path

# Posibles alias de columnas en ITP (ajusta si conoces los exactos)
CAND_ID = ["industry_id", "industry_code", "industry", "sector_id", "sector_code", "isic", "isic4"]
CAND_NAME = ["industry_descr", "industry_name", "industry_label", "sector_name", "sector_descr"]

def _pick_col(candidates, cols):
    cols_lower = {c.lower(): c for c in cols}
    for cand in candidates:
        if cand.lower() in cols_lower:
            return cols_lower[cand.lower()]
    return None

id_col = _pick_col(CAND_ID, data.columns)
name_col = _pick_col(CAND_NAME, data.columns)

if not id_col or not name_col:
    print("⚠️ No se encontraron columnas de industria en 'data'.")
    print(f"Columnas disponibles: {list(data.columns)}")
    print(
        "Sugerencia: en la celda de carga (procesar_datos_itp), añade las columnas de industria "
        "a USECOLS/DTYPES, por ejemplo: 'industry_id' y 'industry_descr' (o sus equivalentes)."
    )
else:
    # Catálogo único, limpio y ordenado
    industrias = (
        data[[id_col, name_col]]
        .dropna()
        .drop_duplicates()
        .sort_values([id_col, name_col])
        .reset_index(drop=True)
        .rename(columns={id_col: "industry_id", name_col: "industry_descr"})
    )

    # Tipos compactos
    industrias["industry_id"] = industrias["industry_id"].astype("string")
    industrias["industry_descr"] = industrias["industry_descr"].astype("string")

    # Rutas (evitamos espacios en nombres de carpetas)
    base_path = Path.cwd().parent.parent
    target_directory = base_path / "data" / "processed" / "Dependencias_consolidadas"
    target_directory.mkdir(parents=True, exist_ok=True)

    # Salidas
    csv_path = target_directory / "industrias_id_nombre.csv"
    pq_path  = target_directory / "industrias_id_nombre.parquet"

    industrias.to_csv(csv_path, sep=";", index=False)
    industrias.to_parquet(pq_path, index=False)

    print(f"✅ Catálogo de industrias generado: {len(industrias)} filas")
    print(f"- CSV: {csv_path}")
    print(f"- Parquet: {pq_path}")




✅ Catálogo de industrias generado: 170 filas
- CSV: c:\Users\Usuario\Documents\Github\Seguridad Economica\data\processed\Dependencias_consolidadas\industrias_id_nombre.csv
- Parquet: c:\Users\Usuario\Documents\Github\Seguridad Economica\data\processed\Dependencias_consolidadas\industrias_id_nombre.parquet


### 📊 Creación de Matrices de Comercio Bilateral Optimizada

Esta función transforma los datos comerciales brutos en un conjunto de **matrices cuadradas alineadas**, 
una por cada industria.

#### 🎯 Qué hace

1. **Valida** que las entradas sean correctas (GroupBy object, países válidos)
2. **Extrae columnas** de exportador, importador y valor de comercio
3. **Para cada industria:**
   - Filtra a solo países en nuestra lista de análisis
   - Crea matriz bilateral: filas=exportadores, columnas=importadores
   - Asegura que todas las matrices sean cuadradas (236×236) y alineadas
4. **Calcula y guarda** totales de importación por país-industria (para auditoría)

#### 📐 Estructura de output
```python
matrices_comercio = {
    "Machinery": DataFrame(236×236),
    "Chemicals": DataFrame(236×236),
    "Electronics": DataFrame(236×236),
    ...
}
```

Donde cada matriz cumple:
- **Índice** = códigos ISO3 de exportadores
- **Columnas** = códigos ISO3 de importadores  
- **Valor [i,j]** = flujo comercial del país i al país j en esa industria
- **Todos los ceros garantizados** = pares no observados = comercio nulo

#### 🔧 Mejoras en esta versión

- ✅ Validación robusta de entrada
- ✅ Mensajes de error claros y accionables
- ✅ Resolución automática de directorios
- ✅ Logging opcional para debugging
- ✅ Comentarios en el código explicando cada paso
- ✅ CSV de referencia para auditoría post-cálculo

#### ⚠️ Notas importantes

- Las matrices se guardan en memoria como **float32** para eficiencia (suficiente para valores de comercio)
- Los totales se guardan como referencia, pero se **recalculan** en la siguiente etapa de análisis
- Las matrices resultantes son la **base para todos los cálculos de dependencia posteriores**

In [5]:
def crear_matriz_comercio_optimizado(
    grouped_data, 
    codigos_paises: List[str],
    target_directory: Path | None = None,
    verbose: bool = False
) -> Dict[str, pd.DataFrame]:
    """
    Crea matrices de comercio bilateral (exportador × importador) para cada industria.
    
    Esta función transforma datos comerciales desagregados en matrices cuadradas alineadas,
    donde cada elemento [i,j] representa el flujo comercial del país i al país j.
    
    Args:
        grouped_data: Resultado de df.groupby('industry_descr'). Debe ser un GroupBy object.
        codigos_paises: Lista de códigos ISO3 a usar como índices de las matrices.
        target_directory: Carpeta donde guardar CSV de totales. Si None, usa path relativa al proyecto.
        verbose: Si True, imprime mensajes de progreso y validación.
    
    Returns:
        dict: {nombre_industria -> DataFrame(index=exportadores, columns=importadores)}
        
    Side effect:
        Guarda CSV comprimido con totales de importación por país-industria para referencia.
    """
    
    # ==================== VALIDACIONES ====================
    
    # Validar que grouped_data sea un GroupBy object
    if not hasattr(grouped_data, '__iter__') or not hasattr(grouped_data, 'obj'):
        raise TypeError(
            "❌ grouped_data debe ser resultado de df.groupby()\n"
            "   Ejemplo: data.groupby('industry_descr')"
        )
    
    # Validar que codigos_paises no esté vacío
    if not codigos_paises or len(codigos_paises) < 2:
        raise ValueError(
            f"❌ codigos_paises debe tener al menos 2 países. Recibido: {len(codigos_paises)}"
        )
    
    if verbose:
        print(f"✓ Entrada validada: {len(codigos_paises)} países")
    
    # ==================== RESOLUCIÓN DE DIRECTORIOS ====================
    
    if target_directory is None:
        try:
            # Si se ejecuta como script
            base_path = Path(__file__).parent
        except NameError:
            # Si se ejecuta en notebook
            base_path = Path.cwd().parent.parent
        
        target_directory = base_path / "data" / "processed" / "totales_comercio_por_pais_ind"
    
    target_directory = Path(target_directory)
    target_directory.mkdir(parents=True, exist_ok=True)
    
    if verbose:
        print(f"✓ Directorio de salida: {target_directory}")
    
    # ==================== BÚSQUEDA DE COLUMNAS REQUERIDAS ====================
    
    cols = grouped_data.obj.columns
    
    # Validar que existan columnas de exportador/importador
    required = {"exporter_iso3", "importer_iso3"}
    missing = required - set(cols)
    if missing:
        raise ValueError(
            f"❌ Faltan columnas requeridas: {missing}\n"
            f"   Disponibles: {list(cols)}"
        )
    
    # Buscar columna de valores de comercio (con prioridad)
    trade_col = None
    for cand in ("trade", "value", "trade_value"):
        if cand in cols:
            trade_col = cand
            break
    
    if trade_col is None:
        # Sugerir alternativas si las hay
        available_value_cols = [c for c in cols if 'trade' in c.lower() or 'value' in c.lower()]
        raise ValueError(
            f"❌ No encontré columna de valores de comercio (buscaba: 'trade', 'value', 'trade_value')\n"
            f"   Candidatos similares: {available_value_cols}\n"
            f"   Todas las columnas: {list(cols)}"
        )
    
    if verbose:
        print(f"✓ Columna de comercio detectada: '{trade_col}'")
    
    # ==================== INICIALIZACIÓN ====================
    
    matrices: Dict[str, pd.DataFrame] = {}
    totales_registros = []
    
    # Pre-calcular template: matriz vacía con índices alineados
    # Todos los países en todas las combinaciones, rellenados con ceros
    template = pd.DataFrame(
        0.0, 
        index=codigos_paises, 
        columns=codigos_paises, 
        dtype=np.float32
    )
    
    # ==================== PROCESAMIENTO POR INDUSTRIA ====================
    
    for industry, group in grouped_data:
        # PASO 1: Filtrar a solo los países en nuestra lista
        # Esto evita que haya países fuera de scope en las matrices
        g = group[
            group["exporter_iso3"].isin(codigos_paises) &
            group["importer_iso3"].isin(codigos_paises)
        ]
        
        # CASO: Industria sin datos válidos
        if g.empty:
            # Guardar matriz vacía (todos ceros) para consistencia
            matrices[industry] = template.copy()
            
            # Registrar totales como cero para todos los países en esta industria
            totales_registros.extend(
                {
                    "pais": p, 
                    "industria": industry, 
                    "valor_importado": 0.0
                }
                for p in codigos_paises
            )
            continue
        
        # CASO: Industria con datos
        # PASO 2: Crear matriz bilateral mediante pivot
        # - index: exportadores (origen)
        # - columns: importadores (destino)
        # - values: suma de comercio (en caso de múltiples registros por par)
        mat = (
            g.pivot_table(
                index="exporter_iso3",
                columns="importer_iso3",
                values=trade_col,
                aggfunc="sum",              # Si hay duplicados, sumarlos
                fill_value=0.0,             # Pares no observados → 0
                observed=False              # Incluir todas las combinaciones
            )
            # PASO 3: Alinear a matriz cuadrada completa
            # Asegura que todos los países estén presentes, aunque no tengan comercio
            .reindex(index=codigos_paises, columns=codigos_paises, fill_value=0.0)
            .astype(np.float32)             # Reducir memoria: float32 suficiente
        )
        
        matrices[industry] = mat
        
        # PASO 4: Calcular totales de importación por país en esta industria
        # .sum(axis=0) suma por columna = suma de todos los exportadores hacia cada importador
        totales = mat.sum(axis=0)
        totales_registros.extend(
            {
                "pais": pais, 
                "industria": industry, 
                "valor_importado": float(totales[pais])
            }
            for pais in codigos_paises
        )
    
    # ==================== GUARDADO DE REFERENCIA ====================
    
    # Convertir a DataFrame y guardar como referencia
    # (se recalcula en calculate_all_dependencies, pero útil para validación/auditoría)
    df_totales = pd.DataFrame(totales_registros)
    df_path = target_directory / f"totales_comercio_por_pais_industria{anio}.csv"
    df_totales.to_csv(df_path, index=False, sep=";")
    
    if verbose:
        print(f"✓ Totales guardados en: {df_path}")
        print(f"✓ Matrices creadas: {len(matrices)} industrias")
    
    return matrices



# Define la lista de códigos de países
codigos_paises = sorted(data['importer_iso3'].unique().tolist())

# Llama a la función optimizada (mismo API esperado)
matrices_comercio = crear_matriz_comercio_optimizado(
    data.groupby('industry_descr'),
    codigos_paises=codigos_paises,
    # target_directory=Path("...")  # opcional
)

def eliminar_filas_columnas_cero(df: pd.DataFrame, threshold_pct: float = 0.005) -> pd.DataFrame:
    """Filtra una matriz aplicando umbral relativo por país importador (columna)."""
    # Umbral por columna: total_col * pct
    col_totals = df.sum(axis=0)
    thresholds = col_totals * float(threshold_pct)

    # Aplicar máscara vectorizada (alineación por columnas)
    df_filtered = df.where(df >= thresholds, 0.0)

    # Países con filas y columnas a cero tras umbral
    zero_rows = df_filtered.index[df_filtered.sum(axis=1) == 0.0]
    zero_cols = df_filtered.columns[df_filtered.sum(axis=0) == 0.0]
    to_drop = list(set(zero_rows) & set(zero_cols))

    return df_filtered.drop(index=to_drop, columns=to_drop)

# 🔗 Cálculo de Dependencias Comerciales: Directas e Indirectas

Esta celda implementa el **núcleo de la metodología PIVI**: calcula cómo cada país depende de otros para cada par bilateral (importador → exportador), considerando no solo el comercio directo sino también las **cadenas de intermediarios** que canalizan el flujo comercial.

#### 🎯 Qué hace

Transforma una matriz de comercio bilateral en un conjunto completo de **métricas de dependencia**:
- **Dependencia directa (L=1):** flujo directo importador ← exportador
- **Dependencia indirecta (L≥2):** flujo canalizado a través de intermediarios
- **Caminos significativos:** rutas concretas que contribuyen a la dependencia
- **Intermediarios críticos:** países que actúan como "hubs" en el comercio

#### 🏗️ Estructura de funciones

**process_country_pair()** - Calcula dependencia para UN par (importador i, exportador j):
- Dependencia directa: X[j,i] / importaciones_totales[i]
- Crea matriz de transición T: normaliza flujos por importador
- L=2 (vectorizado): suma de todos los intermediarios con una operación matricial
- L≥3 (combinations): enumeración de caminos más largos con poda por threshold
- Retorna: dependencias_por_longitud + caminos_significativos

**calculate_all_dependencies_parallel()** - Paraleliza process_country_pair() para TODOS los pares:
- GPU (opcional): prepara aceleración si está disponible
- Joblib: distribuye cálculo en múltiples núcleos
- Agrega resultados por importador
- Calcula centralidad de intermediarios
- Retorna: estructura consolidada de resultados

**calculate_all_dependencies()** - Wrapper que decide si usar paralelización:
- Valida tamaño del problema
- Llama a calculate_all_dependencies_parallel() con parámetros óptimos

#### 📊 Output principal
```python
results = {
    'dependencies': [  # Lista de dicts con cada relación bilateral
        {
            'importador': 'ESP',
            'exportador': 'CHN',
            'dependencia_total': 0.794,
            'dependencia_directa': 0.598,
            'dependencia_indirecta': 0.196,
            'longitud_optima': 3,
            'dependencias_por_longitud': {1: 0.598, 2: 0.162, 3: 0.034},
            'trade_value': 1912.42
        },
        ...  # uno por cada par i≠j (236² - 236 = 54,756 pares)
    ],
    
    'top_dependencies': [  # Top 90 pares más dependientes
        (importador, exportador, DD, DI, DT, longitud), ...
    ],
    
    'critical_paths': [  # Rutas específicas con fuerza > threshold
        {
            'exportador': 'CHN',
            'importador': 'ESP',
            'intermediarios': ['DEU', 'HUN'],
            'fuerza': 0.0039,
            'longitud': 3
        }, ...
    ],
    
    'intermediary_centrality': [  # Países rankeados como intermediarios
        ('DEU', freq=145, strength=23.4, score=0.87),
        ('CHN', freq=132, strength=18.2, score=0.81),
        ...
    ],
    
    'critical_intermediaries': {  # Rutas por cada par bilateral
        'CHN->ESP': [path_dict, path_dict, ...],
        ...
    }
}
```

#### ⚙️ Parámetros clave

| Parámetro | Default | Significado |
|-----------|---------|-------------|
| `max_possible_length` | 3 | Longitud máxima de cadenas a considerar (L_max) |
| `convergence_threshold` | 0.01 | Umbral para detener si incremento < 1% |
| `path_strength_threshold` | 0.001 | Fuerza mínima de un camino para registrarlo |
| `n_jobs` | CPU_count | Número de cores para paralelización |
| `use_gpu` | True | Intenta usar GPU si disponible |

#### 🔢 Complejidad computacional

- **Pares evaluados:** n(n-1) donde n=236 países → ~55k pares
- **Caminos L=2:** O(n) combinaciones por par (rápido, vectorizado)
- **Caminos L≥3:** O(C(n,L-1)) combinaciones → exponencial pero poda por `path_strength_threshold`
- **Tiempo típico:** ~10-30 min para 170 industrias en CPU con paralelización

#### 📈 Matriz de Transición T

Punto crucial del algoritmo:
```
T[exportador, importador] = Comercio(exp→imp) / Importaciones_totales(imp)
```

**Interpretación:** $$T[j,i]$$ representa la "probabilidad" de que una unidad importada por país i provenga del país j, considerando la composición actual del mercado de i. Esto permite calcular cadenas: si j abastece a k y k abastece a i, entonces la cadena es $$T[j,k] × T[k,i]$$.

#### 🧮 Fórmulas de dependencia

**Dependencia Directa (L=1):**
$$DD_{i \leftarrow j} = \frac{X_{j \to i}}{\sum_k X_{k \to i}}$$

**Dependencia Indirecta (L=2):**
$$DI^{(2)}_{i \leftarrow j} = \sum_k T_{j,k} \cdot T_{k,i}$$ 

donde 
$$k ≠ i, j$$

**Dependencia Indirecta (L=3+):**
$$DI^{(L)}_{i \leftarrow j} = \sum_{\text{caminos}} T_{j,c_1} \cdot T_{c_1,c_2} \cdot \ldots \cdot T_{c_{L-2},i}$$

**Dependencia Total:**
$$DT_{i \leftarrow j} = DD_{i \leftarrow j} + \sum_{L=2}^{L_{max}} DI^{(L)}_{i \leftarrow j}$$

#### 🎯 Intermediarios críticos

Para cada país, se calcula:
- **Frecuencia:** cuántas cadenas lo incluyen
- **Fuerza:** suma ponderada de las fuerzas de esas cadenas
- **Score de centralidad:** 0.4 × freq_normalizada + 0.6 × fuerza_normalizada

Un país con alta centralidad es un "cuello de botella" clave en el comercio global.

#### ⚠️ Notas importantes

- El algoritmo es **exhaustivo pero poda intelligentemente**: solo explora caminos con fuerza > `path_strength_threshold`
- La convergencia típicamente se alcanza en L=2 o L=3, rara vez necesita L=4+
- Los resultados son **determinísticos** y **reproducibles**
- Está paralelizado para máximo rendimiento, pero puede ejecutarse secuencial si se necesita debugging



In [6]:
def calculate_path_dependency(X_clean, path, denominators):
    """Calcula la dependencia de un camino específico."""
    fuerza_camino = 1.0
    for k in range(len(path) - 1):
        if denominators[path[k+1]] > 0:
            fuerza_camino *= X_clean[path[k], path[k+1]] / denominators[path[k+1]]
        else:
            fuerza_camino = 0  # Si el denominador es cero, la fuerza del camino es cero
    return fuerza_camino

def calculate_intermediary_centrality(intermediary_frequency, intermediary_strength, country_names):
    """
    Calcula métricas de centralidad para intermediarios.
    
    Esta función es la misma que la original.
    """
    # Implementación existente
    centrality = []
    
    # Normalizar
    max_freq = max(intermediary_frequency.values()) if intermediary_frequency.values() else 1
    max_strength = max(intermediary_strength.values()) if intermediary_strength.values() else 1
    
    for country in country_names:
        norm_freq = intermediary_frequency[country] / max_freq if max_freq > 0 else 0
        norm_strength = intermediary_strength[country] / max_strength if max_strength > 0 else 0
        
        combined_score = 0.4 * norm_freq + 0.6 * norm_strength
        
        centrality.append((country, intermediary_frequency[country], 
                          intermediary_strength[country], combined_score))
    
    centrality.sort(key=lambda x: x[3], reverse=True)
    return centrality


from itertools import combinations
import numpy as np

def process_country_pair(i, j, X_clean, denominators, country_names,
                         max_possible_length, convergence_threshold,
                         path_strength_threshold, T):
    """
    Exactamente la misma definición que antes (combinations), pero:
    - Usa T (matriz de transición) precalculada.
    - Vectoriza la longitud 2.
    - Mantiene mismos retornos y campos.
    """
    n = X_clean.shape[0]

    # Dependencia directa (igual que antes)
    trade_value = X_clean[j, i]
    direct_dependency = (X_clean[j, i] / denominators[i]) if denominators[i] > 0 else 0.0

    dependencies_by_length = {1: direct_dependency}
    significant_paths = []
    current_total = direct_dependency
    indirect_total = 0.0
    length = 1

    # Candidatos de intermediarios (igual que antes)
    middle = [k for k in range(n) if k != i and k != j]

    # ---- L = 2 (vectorizado, EXACTO) ----
    if max_possible_length >= 2 and middle:
        row_j = T[j, :]
        col_i = T[:, i]
        mask = np.ones(n, dtype=bool)
        mask[[i, j]] = False

        # suma exacta: sum_k T[j,k]*T[k,i], k!=i,j
        di2 = np.dot(row_j[mask], col_i[mask])
        dependencies_by_length[2] = float(di2)
        indirect_total += float(di2)
        current_total = direct_dependency + indirect_total
        length = 2

        # caminos significativos para L=2 (mismo umbral)
        if path_strength_threshold > 0:
            ks = np.nonzero(row_j * col_i > path_strength_threshold)[0]
            ks = [k for k in ks if k != i and k != j]
            if ks:
                for k in ks:
                    significant_paths.append({
                        'exportador': country_names[j],
                        'importador': country_names[i],
                        'intermediarios': [country_names[k]],
                        'fuerza': float(row_j[k] * col_i[k]),
                        'longitud': 2
                    })

        # criterio de convergencia
        if abs(current_total - direct_dependency) < convergence_threshold or max_possible_length == 2:
            # ordenar y devolver
            significant_paths.sort(key=lambda x: x['fuerza'], reverse=True)
            pair_key = f"{country_names[j]}->{country_names[i]}"
            result = {
                'importador': country_names[i],
                'exportador': country_names[j],
                'trade_value': trade_value,
                'dependencia_directa': direct_dependency,
                'dependencia_indirecta': indirect_total,
                'dependencia_total': direct_dependency + indirect_total,
                'longitud_optima': 2,
                'dependencias_por_longitud': dependencies_by_length
            }
            return {
                'pair_key': pair_key,
                'result': result,
                'top_dependency': (country_names[i], country_names[j],
                                   direct_dependency, indirect_total,
                                   direct_dependency + indirect_total, 2),
                'significant_paths': significant_paths,
                'length_converged': 2
            }

    # ---- L >= 3 (exacto con combinations, pero usando T) ----
    for L in range(3, max_possible_length + 1):
        DI_ij_L = 0.0
        for interms in combinations(middle, L - 1):
            path = (j,) + interms + (i,)
            # producto exacto de T a lo largo del camino
            prod = 1.0
            for a, b in zip(path[:-1], path[1:]):
                w = T[a, b]
                if w == 0.0:
                    prod = 0.0
                    break
                prod *= w

            DI_ij_L += prod

            if prod > path_strength_threshold:
                significant_paths.append({
                    'exportador': country_names[j],
                    'importador': country_names[i],
                    'intermediarios': [country_names[x] for x in interms],
                    'fuerza': float(prod),
                    'longitud': L
                })

        dependencies_by_length[L] = float(DI_ij_L)
        indirect_total += float(DI_ij_L)

        prev_total = current_total
        current_total = direct_dependency + indirect_total
        length = L

        if L > 1 and abs(current_total - prev_total) < convergence_threshold:
            break

    # Final
    significant_paths.sort(key=lambda x: x['fuerza'], reverse=True)
    total_dependency = direct_dependency + indirect_total
    pair_key = f"{country_names[j]}->{country_names[i]}"
    result = {
        'importador': country_names[i],
        'exportador': country_names[j],
        'trade_value': trade_value,
        'dependencia_directa': direct_dependency,
        'dependencia_indirecta': indirect_total,
        'dependencia_total': total_dependency,
        'longitud_optima': length,
        'dependencias_por_longitud': dependencies_by_length
    }
    return {
        'pair_key': pair_key,
        'result': result,
        'top_dependency': (country_names[i], country_names[j],
                           direct_dependency, indirect_total, total_dependency, length),
        'significant_paths': significant_paths,
        'length_converged': length if length > 1 else 0
    }




def calculate_all_dependencies_parallel(X, country_names=None, convergence_threshold=0.01, 
                                       max_possible_length=3, 
                                       path_strength_threshold=0.001, n_jobs=None, use_gpu=True, 
                                       debug_mode=False):
    """
    Versión paralelizada del cálculo de dependencias que mantiene EXACTAMENTE
    la misma salida que la versión original.
    
    El parámetro debug_mode permite verificar que el número de dependencias
    coincida con la versión original.
    """
    """
    Versión paralelizada del cálculo de dependencias.
    
    Parameters adicionales:
    -----------------------
    n_jobs : int, opcional
        Número de trabajos paralelos. Si es None, usa todos los núcleos disponibles.
    use_gpu : bool, default=True
        Si se debe intentar usar GPU para acelerar algunos cálculos.
    """
    n = X.shape[0]

    if country_names is None:
        country_names = [f"País {i}" for i in range(n)]

    if len(country_names) != n:
        raise ValueError(f"La longitud de country_names ({len(country_names)}) no coincide con la dimensión de X ({n})")

    # Configurar paralelización
    if n_jobs is None:
        n_jobs = multiprocessing.cpu_count()
    
    # Verificar disponibilidad de GPU
    gpu_available = torch.cuda.is_available() and use_gpu
    X_clean = X

    denom = X_clean.sum(axis=0, dtype=np.float64)
    denom[denom == 0.0] = np.inf
    T = (X_clean / denom).astype(np.float64, copy=False)

    denominators = np.sum(X, axis=0)

    # Acelerar cálculos directos con GPU si está disponible
    if gpu_available:
        # Transferir datos a GPU
        X_gpu = torch.tensor(X_clean, device='cuda', dtype=torch.float32)
        denom_gpu = torch.tensor(denominators, device='cuda', dtype=torch.float32)
        
        # Calcular dependencias directas en forma vectorizada
        direct_deps = torch.zeros_like(X_gpu)
        for i in range(n):
            # Evitar división por cero
            if denom_gpu[i] > 0:
                direct_deps[:, i] = X_gpu[:, i] / denom_gpu[i]
        
        # Transferir resultados de vuelta a CPU
        direct_dependencies = direct_deps.cpu().numpy()
        
        # Usar estas dependencias directas precalculadas en el procesamiento posterior
        # (Aunque en esta implementación seguimos calculándolas en process_country_pair para
        # mantener cambios mínimos en el código)

    # Estructura de resultados extendida
    results = {
        'dependencies': [],
        'top_dependencies': [],
        'avg_dependencies': {},
        'length_distribution': np.zeros(max_possible_length),
        'critical_intermediaries': {},     # Intermediarios críticos por relación
        'intermediary_frequency': {},      # Frecuencia de países como intermediarios
        'critical_paths': [],              # Rutas críticas completas
        'intermediary_strength': {}        # Fuerza de cada país como intermediario
    }
    
    # Inicializar contadores para intermediarios
    for country in country_names:
        results['intermediary_frequency'][country] = 0
        results['intermediary_strength'][country] = 0.0

    # Preparar pares de países para procesamiento paralelo 
    # Mantenemos la misma estructura de iteración del código original
    # Primero por importador (i) y luego por exportador (j)
        # Preparar pares de países (sin tqdm)
    country_pairs = [(i, j) for i in range(n) for j in range(n) if i != j]

    # Procesar pares de países en paralelo (sin barra de progreso)
    with Parallel(n_jobs=n_jobs) as parallel:
        pair_results = parallel(
            delayed(process_country_pair)(
                i, j, X_clean, denom, country_names, 
                max_possible_length, convergence_threshold, path_strength_threshold,
                T  # <-- NUEVO ARGUMENTO
            )
            for i, j in country_pairs
        )


        
    # Agrupar resultados por país importador
    results_by_importer = {}
    for res in pair_results:
        importer = res['result']['importador']
        if importer not in results_by_importer:
            results_by_importer[importer] = []
        results_by_importer[importer].append(res)
    
    # Recolectar critical paths de todos los pares para ordenarlos después
    all_critical_paths = []
    
    # Procesar los resultados manteniendo el mismo orden que el código original
    for i in range(n):
        importer = country_names[i]
        total_dep = 0.0
        num_deps = 0
        
        if importer in results_by_importer:
            for res in results_by_importer[importer]:
                # Agregar a dependencies
                results['dependencies'].append(res['result'])
                
                # Agregar a top_dependencies
                results['top_dependencies'].append(res['top_dependency'])
                
                # Actualizar critical_intermediaries
                results['critical_intermediaries'][res['pair_key']] = res['significant_paths']
                
                # Recolectar critical paths
                all_critical_paths.extend(res['significant_paths'])
                
                # Actualizar length_distribution si convergió
                if res['length_converged'] > 1:
                    results['length_distribution'][res['length_converged'] - 1] += 1
                
                # Actualizar dependencia promedio
                total_dep += res['result']['dependencia_total']
                num_deps += 1
                
                # Actualizar estadísticas de intermediarios
                for path in res['significant_paths']:
                    for idx, interm in enumerate(path['intermediarios']):
                        # Incrementar frecuencia
                        results['intermediary_frequency'][interm] += 1
                        
                        # Incrementar fuerza ponderada
                        weight_factor = 1.0 / (idx + 1)
                        results['intermediary_strength'][interm] += path['fuerza'] * weight_factor
        
        # Guardar dependencia promedio para este importador
        results['avg_dependencies'][importer] = total_dep / num_deps if num_deps > 0 else 0
    
    # Añadir y ordenar los critical paths (igual que el original)
    results['critical_paths'] = all_critical_paths
    
    # Ordenar top dependencies
    results['top_dependencies'].sort(key=lambda x: x[4], reverse=True)
    results['top_dependencies'] = results['top_dependencies'][:90]
    
    # Ordenar critical_paths por fuerza
    results['critical_paths'].sort(key=lambda x: x['fuerza'], reverse=True)
    
    # Calcular métricas de centralidad para intermediarios
    results['intermediary_centrality'] = calculate_intermediary_centrality(
        results['intermediary_frequency'], 
        results['intermediary_strength'],
        country_names
    )
    
    return results

# Para mantener compatibilidad, redefinimos la función original
# para que utilice la versión paralelizada
def calculate_all_dependencies(X, country_names=None, convergence_threshold=0.01, 
                              max_possible_length=3, threshold_pct=0.01, 
                              path_strength_threshold=0.001):
    """
    Calcula todas las dependencias entre países con análisis de intermediarios críticos.
    
    Esta función mantiene EXACTAMENTE la misma firma y resultados que la original,
    pero utiliza internamente paralelización y GPU para acelerar los cálculos.
    
    Parameters:
    -----------
    X : numpy.ndarray
        Matriz de comercio
    country_names : list, opcional
        Nombres de los países
    convergence_threshold : float, default=0.01
        Umbral para determinar la convergencia
    max_possible_length : int, default=5
        Longitud máxima de caminos a considerar
    threshold_pct : float, default=0.01
        Umbral para filtrar valores de comercio insignificantes (porcentaje)
    path_strength_threshold : float, default=0.001
        Umbral mínimo para considerar una ruta como significativa
        
    Returns:
    --------
    dict
        Diccionario con todos los resultados del análisis
    """
    # Determinar si usar paralelización basado en el tamaño del problema
    use_parallel = X.shape[0] > 5  # Para matrices muy pequeñas no vale la pena paralelizar
    
    # Verificar disponibilidad de GPU
    use_gpu = torch.cuda.is_available()
    
    # Aquí agregas el segundo log
    #print(f"Usando paralelización: {use_parallel}, GPU disponible: {use_gpu}")
    #if use_gpu:
    #    print(f"GPU en uso: {torch.cuda.get_device_name(0)}")
    
    # Configurar número de trabajos para paralelización
    n_countries = X.shape[0]
    n_jobs = min(multiprocessing.cpu_count(), n_countries)  # Limitar al número de países
    
    
    if use_parallel:
        # Usar la versión paralelizada
        return calculate_all_dependencies_parallel(
            X, country_names, convergence_threshold, 
            max_possible_length, path_strength_threshold,
            n_jobs=n_jobs, use_gpu=use_gpu, debug_mode=False
        )
    else:
        # Para matrices muy pequeñas, usar un solo proceso
        return calculate_all_dependencies_parallel(
            X, country_names, convergence_threshold, 
            max_possible_length, path_strength_threshold,
            n_jobs=1, use_gpu=use_gpu, debug_mode=False
        )


## Generar y guardar resultados de dependencias

In [28]:
# Filtrar para obtener solo las primeras 5 industrias
matrices_comercio = {k: matrices_comercio[k] for k in list(matrices_comercio.keys())[:50]}

# Verificación: imprime las primeras 5 claves
print("Primeras 5 industrias en matrices_comercio:", list(matrices_comercio.keys()))


Primeras 5 industrias en matrices_comercio: ['154 Manufacturing services on physical inputs', '155 Maintenance and repair services n.i.e.', '156 Transport', '157 Travel', '158 Construction', '159 Insurance and pension services', '160 Financial services', '161 Charges for use of intellectual property', '162 Telecom, computer, information services', '163 Other business services', '164 Heritage and recreational services', '165 Health services', '166 Education services', '167 Government goods and services n.i.e.', '168 Services not allocated', '169 Trade-related services', '170 Other personal services', 'Accumulators primary cells and batteries', 'Agricultural and forestry machinery', 'Aircraft and spacecraft', 'Animal feed ingredients and pet foods', 'Articles of concrete cement and plaster', 'Automobile bodies trailers & semi-trailers', 'Bakery products', 'Basic chemicals except fertilizers', 'Basic iron and steel', 'Basic precious and non-ferrous metals', 'Bearings gears gearing & drivi

In [7]:
# Diccionario para guardar todos los resultados
all_results = {}

# Validación inicial
if not matrices_comercio:
    raise ValueError("❌ matrices_comercio está vacío. Verifica crear_matriz_comercio_optimizado()")

total_industrias = len(matrices_comercio)
completadas = 0
saltadas = []
print(f"Procesando {total_industrias} industrias...")

# Procesar cada industria
for industry, mat in matrices_comercio.items():
    try:
        # Limpiar matriz
        mat_clean = eliminar_filas_columnas_cero(mat)
        
        # Si la matriz limpia está vacía o es muy pequeña, continuamos con la siguiente
        if mat_clean.shape[0] < 2:
            saltadas.append(industry)
            continue
        
        # Convertir a numpy array y obtener nombres de países
        X = mat_clean.values
        country_names = mat_clean.columns.tolist()

        # Calcular dependencias
        results = calculate_all_dependencies(X, country_names)
        
        # Guardar resultados
        all_results[industry] = {
            'results': results,
            'country_names': country_names,
            'matrix_shape': mat_clean.shape
        }

        # Actualizar contador y mostrarlo en la misma línea
        completadas += 1
        print(f"✓ Completadas: {completadas}/{total_industrias}", end="\r", flush=True)
    
    except Exception as e:
        print(f"\n⚠️ Error en industria '{industry}': {str(e)}")
        continue

# Resumen final
print(f"\n{'='*80}")
print(f"✅ Procesamiento completado: {completadas}/{total_industrias} industrias")
if saltadas:
    print(f"⚠️ Industrias saltadas ({len(saltadas)}): {', '.join(saltadas[:5])}{'...' if len(saltadas) > 5 else ''}")
print(f"{'='*80}\n")


Procesando 170 industrias...
✓ Completadas: 170/170
✅ Procesamiento completado: 170/170 industrias



## Generar y guadar datos de intermediarios Críticos

In [8]:
def get_top_intermediaries(results, top_n=10):
    """
    Obtiene los países más importantes como intermediarios.
    
    Parameters:
    -----------
    results : dict
        Resultados del análisis con intermediarios
    top_n : int
        Número de países a mostrar
        
    Returns:
    --------
    list
        Lista de tuplas (país, score, frecuencia, fuerza) ordenadas por importancia
    """
    # La estructura de intermediary_centrality ha cambiado en la nueva implementación
    # Ahora es una lista de tuplas (país, frecuencia, fuerza, score_combinado)
    if isinstance(results['intermediary_centrality'], list):
        # Nueva implementación: ya está ordenada por score combinado
        return results['intermediary_centrality'][:top_n]
    else:
        # Implementación anterior (por compatibilidad)
        centrality_scores = [(country, metrics) 
                            for country, metrics in results['intermediary_centrality'].items()]
        centrality_scores.sort(key=lambda x: x[1]['centrality_score'], reverse=True)
        return centrality_scores[:top_n]

def analyze_country_intermediary_role(results, country):
    """
    Analiza el papel de un país específico como intermediario.
    
    Parameters:
    -----------
    results : dict
        Resultados del análisis con intermediarios
    country : str
        País a analizar
        
    Returns:
    --------
    dict
        Información detallada sobre el rol del país como intermediario
    """
    # Verificar si el país existe en los resultados
    if isinstance(results['intermediary_centrality'], list):
        # Nueva implementación - lista de tuplas
        country_entry = next((entry for entry in results['intermediary_centrality'] 
                             if entry[0] == country), None)
        
        if country_entry is None:
            return {"error": f"El país {country} no está en los resultados"}
            
        # Extraer métricas
        centrality = {
            "frequency": country_entry[1],
            "strength": country_entry[2],
            "centrality_score": country_entry[3]
        }
    else:
        # Implementación anterior
        if country not in results['intermediary_centrality']:
            return {"error": f"El país {country} no está en los resultados"}
        centrality = results['intermediary_centrality'][country]
    
    # Encontrar las rutas más importantes donde este país actúa como intermediario
    top_paths = []
    for path in results['critical_paths']:
        if country in path['intermediarios']:
            top_paths.append(path)
    
    # Ordenar por fuerza del camino
    top_paths.sort(key=lambda x: x['fuerza'], reverse=True)
    
    # Filtrar los 10 caminos más importantes
    top_paths = top_paths[:10]
    
    return {
        "metrics": centrality,
        "top_paths": top_paths,
        "role_summary": {
            "total_paths": len(top_paths),
            "average_path_strength": sum(p['fuerza'] for p in top_paths) / len(top_paths) if top_paths else 0,
            "max_path_strength": max(p['fuerza'] for p in top_paths) if top_paths else 0,
            "unique_exporters": len(set(p['exportador'] for p in top_paths)),
            "unique_importers": len(set(p['importador'] for p in top_paths))
        }
    }

def summarize_country_dependencies(results, country, top_n=5):
    """
    Genera un resumen de las dependencias comerciales de un país específico.
    
    Parameters:
    -----------
    results : dict
        Resultados del análisis de dependencias
    country : str
        País a analizar
    top_n : int
        Número de dependencias principales a mostrar
        
    Returns:
    --------
    dict
        Resumen de dependencias del país
    """
    # Dependencias como importador (otros países exportan a este país)
    import_dependencies = []
    for dep in results['dependencies']:
        if dep['importador'] == country:
            import_dependencies.append(dep)
    
    # Ordenar por dependencia total
    import_dependencies.sort(key=lambda x: x['dependencia_total'], reverse=True)
    
    # Dependencias como exportador (este país exporta a otros)
    export_dependencies = []
    for dep in results['dependencies']:
        if dep['exportador'] == country:
            export_dependencies.append(dep)
    
    # Ordenar por dependencia total
    export_dependencies.sort(key=lambda x: x['dependencia_total'], reverse=True)
    
    # Calcular dependencia promedio
    avg_dependency = results['avg_dependencies'].get(country, 0)
    
    # Analizar el papel como intermediario
    intermediary_role = None
    if isinstance(results['intermediary_centrality'], list):
        # Nueva implementación
        intermediary_info = next((x for x in results['intermediary_centrality'] if x[0] == country), None)
        if intermediary_info:
            intermediary_role = {
                "frequency": intermediary_info[1],
                "strength": intermediary_info[2],
                "centrality_score": intermediary_info[3],
                "rank": next((i+1 for i, x in enumerate(results['intermediary_centrality']) 
                             if x[0] == country), None)
            }
    else:
        # Implementación anterior
        if country in results['intermediary_centrality']:
            intermediary_role = results['intermediary_centrality'][country]
            # Calcular rango
            countries_sorted = sorted(results['intermediary_centrality'].keys(), 
                                     key=lambda x: results['intermediary_centrality'][x]['centrality_score'],
                                     reverse=True)
            intermediary_role["rank"] = countries_sorted.index(country) + 1
    
    return {
        "country": country,
        "avg_dependency": avg_dependency,
        "top_import_dependencies": import_dependencies[:top_n],
        "top_export_dependencies": export_dependencies[:top_n],
        "total_import_dependencies": len(import_dependencies),
        "total_export_dependencies": len(export_dependencies),
        "intermediary_role": intermediary_role
    }

def identify_critical_trade_relationships(results, threshold=0.7, min_paths=3):
    """
    Identifica relaciones comerciales críticas con alta dependencia.
    
    Parameters:
    -----------
    results : dict
        Resultados del análisis de dependencias
    threshold : float
        Umbral de dependencia para considerar una relación como crítica
    min_paths : int
        Número mínimo de caminos alternativos para evitar criticidad
        
    Returns:
    --------
    list
        Lista de relaciones críticas
    """
    critical_relationships = []
    
    # Analizar cada dependencia
    for dep in results['dependencies']:
        if dep['dependencia_total'] >= threshold:
            # Buscar caminos alternativos
            pair_key = f"{dep['exportador']}->{dep['importador']}"
            alternative_paths = []
            
            if pair_key in results['critical_intermediaries']:
                alternative_paths = results['critical_intermediaries'][pair_key]
            
            # Si hay pocos caminos alternativos, es una relación crítica
            if len(alternative_paths) < min_paths:
                critical_relationships.append({
                    "exportador": dep['exportador'],
                    "importador": dep['importador'],
                    "dependencia_total": dep['dependencia_total'],
                    "dependencia_directa": dep['dependencia_directa'],
                    "caminos_alternativos": len(alternative_paths),
                    "criticidad": 1.0 - (len(alternative_paths) / min_paths) 
                                     if min_paths > 0 else 1.0
                })
    
    # Ordenar por criticidad
    critical_relationships.sort(key=lambda x: x['criticidad'], reverse=True)
    
    return critical_relationships

## Creo Dataframes para guardar resultados

In [9]:
import pandas as pd

def create_dependencies_dataframe(all_results):
    """
    Crea un DataFrame con los resultados de dependencias para todas las industrias.
    
    Parameters:
    -----------
    all_results : dict
        Diccionario con los resultados por industria
        
    Returns:
    --------
    pandas.DataFrame
        DataFrame con las columnas: industria, importador, exportador, 
        dependencia_total, dependencia_directa, dependencia_indirecta, longitud_optima
    """
    # Lista para almacenar los datos de todas las industrias
    all_data = []
    
    # Procesar cada industria
    for industry, data in all_results.items():
        # Obtener los resultados de esta industria
        industry_results = data['results']['dependencies']
        
        # Añadir cada fila de resultados
        for result in industry_results:
            row = {
                'industria': industry,
                'importador': result['importador'],
                'exportador': result['exportador'],
                'dependencia_total': result['dependencia_total'],
                'dependencia_directa': result['dependencia_directa'],
                'dependencia_indirecta': result['dependencia_indirecta'],
                'trade_value': result['trade_value'],
                'longitud_optima': result['longitud_optima']
            }
            all_data.append(row)
    
    # Crear el DataFrame
    df = pd.DataFrame(all_data)
    
    # Ordenar el DataFrame
    df = df.sort_values(['industria', 'dependencia_total'], ascending=[True, False])
    
    return df

# Ejemplo de uso:
df = create_dependencies_dataframe(all_results)

# Definir el nuevo mapeo de nombres de columnas
nuevo_nombres = {
    'industria': 'industry',
    'importador': 'dependent_country',
    'exportador': 'supplier_country',
    'dependencia_total': 'dependency_value',
    'dependencia_directa': 'direct_dependency',
    'dependencia_indirecta': 'indirect_dependency',
    'trade_value': 'trade_value',
    'longitud_optima': 'longitud_optima'
}

# Renombrar las columnas
df = df.rename(columns=nuevo_nombres)

# Rutas (evitamos espacios en nombres de carpetas)
base_path = Path.cwd().parent.parent
target_directory = base_path / "data" / "processed" / "dependencias_consolidadas"

ruta_archivo = target_directory / f"dependencias{anio}_borrar.csv.gz"


# Guardar como gzip
with gzip.open(ruta_archivo, 'wt', encoding='utf-8') as f:
    df.to_csv(f, sep=";", index=False)

print(f"DataFrame guardado correctamente en: {ruta_archivo}")

DataFrame guardado correctamente en: c:\Users\Usuario\Documents\Github\Seguridad Economica\data\processed\dependencias_consolidadas\dependencias2022_borrar.csv.gz


# ANÁLISIS

In [10]:
def analizar_dependencia_trazable(
    matrices_comercio, 
    industry: str,
    exportador_iso3: str,
    importador_iso3: str,
    country_names=None,
    convergence_threshold=0.01,
    max_possible_length=3,
    path_strength_threshold=0.001,
    apply_cleanup=True,  # ← NUEVO
    verbose=True
):
    """
    Calcula la dependencia de un par exportador-importador en una industria específica,
    mostrando el RASTRO COMPLETO de cada cálculo paso a paso.
    
    AHORA TAMBIÉN MUESTRA de dónde salen las FUERZAS en dependencias indirectas:
    - Valores brutos de comercio X
    - Denominadores (importaciones totales)
    - Cálculo de cada transición T
    - Producto final
    
    Parameters:
    -----------
    matrices_comercio : dict
        Diccionario de matrices por industria (salida de crear_matriz_comercio_optimizado)
    industry : str
        Nombre de la industria a analizar
    exportador_iso3 : str
        Código ISO3 del país exportador
    importador_iso3 : str
        Código ISO3 del país importador
    country_names : list, opcional
        Lista de nombres de países (si None, usa los índices de la matriz)
    convergence_threshold : float
        Umbral de convergencia
    max_possible_length : int
        Longitud máxima de caminos
    path_strength_threshold : float
        Umbral mínimo de fuerza de camino
    verbose : bool
        Si imprimir el análisis detallado
        
    Returns:
    --------
    dict con toda la información y rastro de cálculos
    """
    
    # VALIDACIONES
    if industry not in matrices_comercio:
        raise ValueError(f"Industria '{industry}' no encontrada. Disponibles: {list(matrices_comercio.keys())[:5]}...")
    
    mat = matrices_comercio[industry]
    
    if apply_cleanup:
        mat = eliminar_filas_columnas_cero(mat)
    
    # Obtener índices de países
    try:
        j_idx = list(mat.index).index(exportador_iso3)
    except ValueError:
        raise ValueError(f"Exportador '{exportador_iso3}' no en industria '{industry}'")
    
    try:
        i_idx = list(mat.columns).index(importador_iso3)
    except ValueError:
        raise ValueError(f"Importador '{importador_iso3}' no en industria '{industry}'")
    
    # Convertir a numpy
    X = mat.values.astype(np.float64)
    pais_names = list(mat.columns)
    n = X.shape[0]
    
    if country_names is None:
        country_names = pais_names
    
    # =====================================================
    # RASTRO DE CÁLCULOS
    # =====================================================
    trace = {
        'industry': industry,
        'exportador': exportador_iso3,
        'importador': importador_iso3,
        'pasos': []
    }
    
    # PASO 1: MATRIZ DE TRANSICIÓN
    # ============================
    denom = X.sum(axis=0, dtype=np.float64)
    denom_safe = denom.copy()
    denom_safe[denom_safe == 0.0] = np.inf
    T = (X / denom_safe).astype(np.float64)
    
    trace['pasos'].append({
        'fase': '1. MATRIZ DE TRANSICIÓN',
        'descripcion': f'T[a,b] = X[a,b] / (suma de importaciones de b)',
        'denominadores': {pais_names[k]: float(denom[k]) for k in range(n)},
        'matriz_T_forma': T.shape
    })
    
    # PASO 2: DEPENDENCIA DIRECTA (L=1)
    # ==================================
    trade_value = X[j_idx, i_idx]
    denom_i = denom[i_idx]
    direct_dependency = (X[j_idx, i_idx] / denom_i) if denom_i > 0 else 0.0
    
    trace['pasos'].append({
        'fase': '2. DEPENDENCIA DIRECTA (L=1)',
        'calculo': f'{exportador_iso3} → {importador_iso3}',
        'valor_comercio': float(trade_value),
        'formula': f'X[{j_idx},{i_idx}] / denom[{i_idx}]',
        'valor_numerador': float(X[j_idx, i_idx]),
        'denominador': float(denom_i),
        'resultado': float(direct_dependency)
    })
    
    total_dependency = direct_dependency
    dependencies_by_length = {1: direct_dependency}
    all_paths = []
    
    # PASO 3: DEPENDENCIA INDIRECTA L=2
    # ==================================
    if max_possible_length >= 2:
        trace_l2 = {
            'fase': '3. DEPENDENCIA INDIRECTA (L=2)',
            'formula': f'sum_k [ T[{j_idx},k] * T[k,{i_idx}] ] para k != {j_idx}, {i_idx}',
            'intermediarios_evaluados': [],
            'suma_total': 0.0
        }
        
        di_l2 = 0.0
        row_j = T[j_idx, :]
        col_i = T[:, i_idx]
        
        for k in range(n):
            if k != j_idx and k != i_idx:
                # PASO A: {exportador} → {intermediario}
                x_jk = X[j_idx, k]
                denom_k = denom[k]
                t_jk = row_j[k]
                
                # PASO B: {intermediario} → {importador}
                x_ki = X[k, i_idx]
                denom_i_paso2 = denom[i_idx]
                t_ki = col_i[k]
                
                # Producto
                producto = t_jk * t_ki
                
                trace_l2['intermediarios_evaluados'].append({
                    'intermediario': pais_names[k],
                    'paso_1_exportador_a_intermediario': {
                        'transicion': f'{exportador_iso3} → {pais_names[k]}',
                        'valor_comercio_X': float(x_jk),
                        'importaciones_totales_de': f'{pais_names[k]}',
                        'denominador': float(denom_k),
                        'T_valor': float(t_jk),
                        'formula': f'X[{j_idx},{k}] / denom[{k}] = {x_jk:.10f} / {denom_k:.10f} = {t_jk:.10f}'
                    },
                    'paso_2_intermediario_a_importador': {
                        'transicion': f'{pais_names[k]} → {importador_iso3}',
                        'valor_comercio_X': float(x_ki),
                        'importaciones_totales_de': f'{importador_iso3}',
                        'denominador': float(denom_i_paso2),
                        'T_valor': float(t_ki),
                        'formula': f'X[{k},{i_idx}] / denom[{i_idx}] = {x_ki:.10f} / {denom_i_paso2:.10f} = {t_ki:.10f}'
                    },
                    'producto_fuerza': {
                        'formula': f'T[{j_idx},{k}] × T[{k},{i_idx}] = {t_jk:.10f} × {t_ki:.10f}',
                        'resultado': float(producto)
                    },
                    'pasa_threshold': producto > path_strength_threshold
                })
                
                di_l2 += producto
                
                if producto > path_strength_threshold:
                    all_paths.append({
                        'camino': f"{exportador_iso3} → {pais_names[k]} → {importador_iso3}",
                        'longitud': 2,
                        'intermediarios': [pais_names[k]],
                        'fuerza': float(producto),
                        'desglose_pasos': [
                            {
                                'de': exportador_iso3,
                                'a': pais_names[k],
                                'valor_X': float(x_jk),
                                'importaciones_totales': float(denom_k),
                                'T_valor': float(t_jk),
                                'formula': f'{x_jk:.10f} / {denom_k:.10f} = {t_jk:.10f}'
                            },
                            {
                                'de': pais_names[k],
                                'a': importador_iso3,
                                'valor_X': float(x_ki),
                                'importaciones_totales': float(denom_i_paso2),
                                'T_valor': float(t_ki),
                                'formula': f'{x_ki:.10f} / {denom_i_paso2:.10f} = {t_ki:.10f}'
                            }
                        ],
                        'pasos_transicion': [
                            (exportador_iso3, pais_names[k], float(t_jk)),
                            (pais_names[k], importador_iso3, float(t_ki))
                        ]
                    })
        
        trace_l2['suma_total'] = float(di_l2)
        trace['pasos'].append(trace_l2)
        
        dependencies_by_length[2] = float(di_l2)
        total_dependency += di_l2
    
    # PASO 4+: DEPENDENCIA INDIRECTA L>=3
    # ====================================
    if max_possible_length >= 3:
        middle = [k for k in range(n) if k != j_idx and k != i_idx]
        
        for L in range(3, max_possible_length + 1):
            trace_lx = {
                'fase': f'4.{L-2}. DEPENDENCIA INDIRECTA (L={L})',
                'longitud': L,
                'num_combinaciones': len(list(combinations(middle, L-1))),
                'caminos_significativos': [],
                'suma_total': 0.0
            }
            
            di_lx = 0.0
            
            for interms in combinations(middle, L-1):
                path_seq = (j_idx,) + interms + (i_idx,)
                
                # Calcular producto a lo largo del camino CON DESGLOSE
                prod = 1.0
                pasos_camino = []
                desglose_pasos = []
                
                for a, b in zip(path_seq[:-1], path_seq[1:]):
                    x_ab = X[a, b]
                    denom_b = denom[b]
                    w = T[a, b]
                    
                    desglose_pasos.append({
                        'de': pais_names[a],
                        'a': pais_names[b],
                        'valor_X': float(x_ab),
                        'importaciones_totales': float(denom_b),
                        'T_valor': float(w),
                        'formula': f'{x_ab:.10f} / {denom_b:.10f} = {w:.10f}'
                    })
                    
                    pasos_camino.append((pais_names[a], pais_names[b], float(w)))
                    if w == 0.0:
                        prod = 0.0
                        break
                    prod *= w
                
                di_lx += prod
                
                if prod > path_strength_threshold:
                    # Construir fórmula del producto
                    formula_prod = ' × '.join([f'{paso["T_valor"]:.10f}' for paso in desglose_pasos])
                    
                    trace_lx['caminos_significativos'].append({
                        'camino': ' → '.join([pais_names[idx] for idx in path_seq]),
                        'intermediarios': [pais_names[idx] for idx in interms],
                        'pasos_desglose': desglose_pasos,
                        'formula_producto': formula_prod,
                        'producto': float(prod)
                    })
                    
                    all_paths.append({
                        'camino': ' → '.join([pais_names[idx] for idx in path_seq]),
                        'longitud': L,
                        'intermediarios': [pais_names[idx] for idx in interms],
                        'fuerza': float(prod),
                        'desglose_pasos': desglose_pasos,
                        'pasos_transicion': pasos_camino
                    })
            
            trace_lx['suma_total'] = float(di_lx)
            trace['pasos'].append(trace_lx)
            
            dependencies_by_length[L] = float(di_lx)
            total_dependency += di_lx
            
            # Criterio de convergencia
            if L > 1 and di_lx < convergence_threshold:
                trace['pasos'].append({
                    'fase': 'CONVERGENCIA',
                    'mensaje': f'L={L}: contribución ({di_lx}) < threshold ({convergence_threshold})',
                    'detenido_en': L
                })
                break
    
    # RESUMEN FINAL
    # =============
    trace['resumen'] = {
        'dependencia_directa': float(direct_dependency),
        'dependencia_indirecta_total': float(total_dependency - direct_dependency),
        'dependencia_total': float(total_dependency),
        'desglose_por_longitud': {k: float(v) for k, v in dependencies_by_length.items()},
        'num_caminos_significativos': len(all_paths),
        'valor_comercio': float(trade_value)
    }
    
    trace['caminos'] = sorted(all_paths, key=lambda x: x['fuerza'], reverse=True)
    
    # IMPRESIÓN FORMATEADA
    # ====================
    if verbose:
        print("\n" + "="*100)
        print(f"ANÁLISIS TRAZABLE DE DEPENDENCIA COMERCIAL CON DESGLOSE DE FUERZAS")
        print("="*100)
        print(f"\n📊 CONTEXTO:")
        print(f"   Industria:       {industry}")
        print(f"   Exportador:      {exportador_iso3}")
        print(f"   Importador:      {importador_iso3}")
        print(f"   Valor comercio:  {trade_value:,.2f}")
        
        print(f"\n{'─'*100}")
        print(f"MATRIZ DE TRANSICIÓN T")
        print(f"{'─'*100}")
        print(f"T[a,b] = (flujo de a hacia b) / (suma total de importaciones de b)")
        print(f"Dimensiones: {T.shape[0]} países × {T.shape[1]} países")
        
        print(f"\n{'─'*100}")
        print(f"CÁLCULO DE DEPENDENCIA")
        print(f"{'─'*100}")
        
        print(f"\n1️⃣  DEPENDENCIA DIRECTA (L=1):")
        print(f"   Fórmula: X[{j_idx},{i_idx}] / Σ(importaciones de {i_idx})")
        print(f"   Numerador:   {X[j_idx, i_idx]:.10f}")
        print(f"   Denominador: {denom_i:.10f}")
        print(f"   ➜ Resultado: {direct_dependency:.10f}")
        
        if max_possible_length >= 2:
            di_l2 = dependencies_by_length.get(2, 0)
            print(f"\n2️⃣  DEPENDENCIA INDIRECTA (L=2):")
            print(f"   Caminos evaluados: {n - 2} (intermediarios posibles)")
            
            # Mostrar top 5 intermediarios CON DESGLOSE
            l2_paths = [p for p in all_paths if p['longitud'] == 2]
            if l2_paths:
                l2_paths_sorted = sorted(l2_paths, key=lambda x: x['fuerza'], reverse=True)
                print(f"   Top intermediarios CON DESGLOSE DE FUERZAS:")
                for pidx, path in enumerate(l2_paths_sorted[:5], 1):
                    inter = path['intermediarios'][0]
                    print(f"\n      {pidx}. {inter} (Fuerza: {path['fuerza']:.10f})")
                    
                    # Paso 1
                    paso1 = path['desglose_pasos'][0]
                    print(f"         └─ PASO 1: {paso1['de']} → {paso1['a']}")
                    print(f"            X = {paso1['valor_X']:.10f}")
                    print(f"            Importaciones totales de {paso1['a']} = {paso1['importaciones_totales']:.10f}")
                    print(f"            T = {paso1['valor_X']:.10f} / {paso1['importaciones_totales']:.10f} = {paso1['T_valor']:.10f}")
                    
                    # Paso 2
                    paso2 = path['desglose_pasos'][1]
                    print(f"         └─ PASO 2: {paso2['de']} → {paso2['a']}")
                    print(f"            X = {paso2['valor_X']:.10f}")
                    print(f"            Importaciones totales de {paso2['a']} = {paso2['importaciones_totales']:.10f}")
                    print(f"            T = {paso2['valor_X']:.10f} / {paso2['importaciones_totales']:.10f} = {paso2['T_valor']:.10f}")
                    
                    # Producto
                    print(f"         └─ FUERZA = {paso1['T_valor']:.10f} × {paso2['T_valor']:.10f} = {path['fuerza']:.10f}")
                
                if len(l2_paths_sorted) > 5:
                    print(f"\n      ... y {len(l2_paths_sorted)-5} más")
            
            print(f"\n   ➜ Suma L=2: {di_l2:.10f}")
        
        if max_possible_length >= 3:
            for L in range(3, max_possible_length + 1):
                di_lx = dependencies_by_length.get(L, 0)
                lx_paths = [p for p in all_paths if p['longitud'] == L]
                print(f"\n{L}️⃣  DEPENDENCIA INDIRECTA (L={L}):")
                print(f"   Combinaciones evaluadas: {len(list(combinations(middle, L-1)))}")
                print(f"   Caminos significativos encontrados: {len(lx_paths)}")
                
                if lx_paths:
                    lx_paths_sorted = sorted(lx_paths, key=lambda x: x['fuerza'], reverse=True)
                    print(f"   Top caminos CON DESGLOSE:")
                    for pidx, path in enumerate(lx_paths_sorted[:3], 1):
                        print(f"\n      {pidx}. {path['camino']}")
                        print(f"         Fuerza: {path['fuerza']:.10f}")
                        
                        for step_idx, paso in enumerate(path['desglose_pasos'], 1):
                            print(f"         Paso {step_idx}: {paso['de']} → {paso['a']}")
                            print(f"            X={paso['valor_X']:.10f}, Importaciones={paso['importaciones_totales']:.10f}")
                            print(f"            T = {paso['valor_X']:.10f} / {paso['importaciones_totales']:.10f} = {paso['T_valor']:.10f}")
                        
                        # Producto final
                        formula_prod = ' × '.join([f'{p["T_valor"]:.10f}' for p in path['desglose_pasos']])
                        print(f"         PRODUCTO = {formula_prod} = {path['fuerza']:.10f}")
                    
                    if len(lx_paths_sorted) > 3:
                        print(f"\n      ... y {len(lx_paths_sorted)-3} más")
                
                print(f"\n   ➜ Suma L={L}: {di_lx:.10f}")
        
        print(f"\n{'─'*100}")
        print(f"RESUMEN FINAL")
        print(f"{'─'*100}")
        indirect = total_dependency - direct_dependency
        print(f"   Dependencia DIRECTA:        {direct_dependency:.10f} ({100*direct_dependency/total_dependency if total_dependency > 0 else 0:.1f}%)")
        print(f"   Dependencia INDIRECTA:      {indirect:.10f} ({100*indirect/total_dependency if total_dependency > 0 else 0:.1f}%)")
        print(f"   ╔═══════════════════════════════════╗")
        print(f"   ║ DEPENDENCIA TOTAL:  {total_dependency:.10f} ║")
        print(f"   ╚═══════════════════════════════════╝")
        
        print(f"\nDesglose por longitud de camino:")
        for L, dep in sorted(dependencies_by_length.items()):
            pct = 100*dep/total_dependency if total_dependency > 0 else 0
            print(f"   L={L}: {dep:.10f} ({pct:.1f}%)")
        
        print(f"\n" + "="*100 + "\n")
    
    return trace

In [16]:
# Analizar una dependencia específica
trace = analizar_dependencia_trazable(
    matrices_comercio,
    industry="Accumulators primary cells and batteries",           # Nombre de la industria
    exportador_iso3="CHN",            # Código ISO3 del exportador
    importador_iso3="ESP",            # Código ISO3 del importador
    max_possible_length=3,
    path_strength_threshold=0.001,
    apply_cleanup=True,  # ← Ahora limpia igual que el original
    verbose=True                      # Imprime análisis detallado
)


ANÁLISIS TRAZABLE DE DEPENDENCIA COMERCIAL CON DESGLOSE DE FUERZAS

📊 CONTEXTO:
   Industria:       Accumulators primary cells and batteries
   Exportador:      CHN
   Importador:      ESP
   Valor comercio:  1,912.42

────────────────────────────────────────────────────────────────────────────────────────────────────
MATRIZ DE TRANSICIÓN T
────────────────────────────────────────────────────────────────────────────────────────────────────
T[a,b] = (flujo de a hacia b) / (suma total de importaciones de b)
Dimensiones: 232 países × 232 países

────────────────────────────────────────────────────────────────────────────────────────────────────
CÁLCULO DE DEPENDENCIA
────────────────────────────────────────────────────────────────────────────────────────────────────

1️⃣  DEPENDENCIA DIRECTA (L=1):
   Fórmula: X[41,67] / Σ(importaciones de 67)
   Numerador:   1912.4212646484
   Denominador: 3196.0346393585
   ➜ Resultado: 0.5983731344

2️⃣  DEPENDENCIA INDIRECTA (L=2):
   Caminos evaluad

# Creamos intermediarios_globales.parquet

In [11]:
# Crear un DataFrame vacío para almacenar los resultados
intermediarios_globales = []

# Recorrer los resultados de cada industria
for industry, data in all_results.items():
    # Obtener los resultados de intermediarios para esta industria
    for country, freq, strength, score in data['results']['intermediary_centrality']:
        intermediarios_globales.append({
            'industry': industry,
            'country': country,
            'frequency_total': freq,
            'strength_total': strength,
            'global_score': score
        })

# Crear el DataFrame
df_intermediarios = pd.DataFrame(intermediarios_globales)

# Guardar el DataFrame como Parquet
output_path = Path.cwd().parent.parent / "data" / "processed" / "dependencias_consolidadas" / "intermediarios_globales.parquet"
df_intermediarios.to_parquet(output_path, index=False)
print(f"✅ intermediarios_globales guardado en: {output_path}")


✅ intermediarios_globales guardado en: c:\Users\Usuario\Documents\Github\Seguridad Economica\data\processed\dependencias_consolidadas\intermediarios_globales.parquet


country_profiles.parquet

In [12]:
# Crear una lista para almacenar los perfiles de los países
country_profiles = []

# Recorrer los resultados de cada industria
for industry, data in all_results.items():
    # Obtener los resultados de dependencias por país
    for country in data['results']['dependencies']:
        # Agregar el perfil del país
        country_profiles.append({
            'industry': industry,
            'country': country['importador'],
            'vulnerability': country.get('dependencia_total', 0),
            'importance': country.get('dependencia_total', 0),
            'num_suppliers': country.get('num_suppliers', 0)
        })

# Crear el DataFrame
df_country_profiles = pd.DataFrame(country_profiles)

# Guardar como Parquet
output_path = Path.cwd().parent.parent / "data" / "processed" / "dependencias_consolidadas" / "country_profiles.parquet"
df_country_profiles.to_parquet(output_path, index=False)
print(f"✅ country_profiles guardado en: {output_path}")


✅ country_profiles guardado en: c:\Users\Usuario\Documents\Github\Seguridad Economica\data\processed\dependencias_consolidadas\country_profiles.parquet


relaciones_criticas.parquet

In [13]:
# Lista para almacenar las relaciones críticas
relaciones_criticas = []

# Recorrer los resultados de cada industria
for industry, data in all_results.items():
    # Obtener las relaciones críticas de esta industria
    crits = identify_critical_trade_relationships(data['results'], threshold=0.7, min_paths=3)
    for c in crits:
        c['industry'] = industry
        relaciones_criticas.append(c)

# Crear el DataFrame
df_relaciones_criticas = pd.DataFrame(relaciones_criticas)

# Guardar como Parquet
output_path = Path.cwd().parent.parent / "data" / "processed" / "dependencias_consolidadas" / "relaciones_criticas.parquet"
df_relaciones_criticas.to_parquet(output_path, index=False)
print(f"✅ relaciones_criticas guardado en: {output_path}")


✅ relaciones_criticas guardado en: c:\Users\Usuario\Documents\Github\Seguridad Economica\data\processed\dependencias_consolidadas\relaciones_criticas.parquet


caminos_significativos.parquet

In [14]:
# Lista para almacenar los caminos significativos
caminos_significativos = []

# Recorrer los resultados de cada industria
for industry, data in all_results.items():
    # Obtener los caminos significativos de esta industria
    for path in data['results']['critical_paths']:
        caminos_significativos.append({
            'industry': industry,
            'exportador': path['exportador'],
            'importador': path['importador'],
            'intermediarios': path['intermediarios'],
            'fuerza': path['fuerza'],
            'longitud': path['longitud']
        })

# Crear el DataFrame
df_caminos_significativos = pd.DataFrame(caminos_significativos)

# Guardar como Parquet
output_path = Path.cwd().parent.parent / "data" / "processed" / "dependencias_consolidadas" / "caminos_significativos.parquet"
df_caminos_significativos.to_parquet(output_path, index=False)
print(f"✅ caminos_significativos guardado en: {output_path}")


✅ caminos_significativos guardado en: c:\Users\Usuario\Documents\Github\Seguridad Economica\data\processed\dependencias_consolidadas\caminos_significativos.parquet


Verificar que all_results esté disponible:

In [15]:
# Guardar all_results como archivo .pkl
import pickle

# Asegúrate de que 'all_results' se haya creado antes de esta celda
pkl_path = Path.cwd().parent.parent / "data" / "processed" / "dependencias_consolidadas" / "all_results.pkl"
with open(pkl_path, "wb") as f:
    pickle.dump(all_results, f)

print(f"✅ all_results guardado en: {pkl_path}")


✅ all_results guardado en: c:\Users\Usuario\Documents\Github\Seguridad Economica\data\processed\dependencias_consolidadas\all_results.pkl
